proposals/0006-react-native-monorepo.md

```diff
@@ -0,0 +1,166 @@
+
+### A IRL scenario of this being a problem: `react-native-codegen` and RN 68
+
+To clarify further how this current structure negatively affects maintaining the codebase, here's a quick example that has happened relatively recently.
+
+`react-native-codegen` is one of the packages shown in the graph as one that is on a `0.0.X` versioning. This meant that with RN `0.68.0`, the version of `react-native-codegen` it depended on was the one the codebase was at the time of cutting the `0.68-stable` branch (`0.0.13`).
+
+After releasing `0.68.0`, it was quickly realised that there was an issue with this version (details are unnecessary) that required a fix; this fix also needed to affect code within the `react-native-codegen` folder.
+
+Here's the catch: in the time between the cut of `0.68-stable` branch and the release of `0.68.0`, work on the main branch kept on going and by when the fix was needed, `react-native-codegen` was on `0.0.16`. This meant that we needed to apply a 0.68 specific fix on `react-native-codegen` in its code shape in the `0.68-stable` branch, and do a release in such a way that it would only get picked up by RN 68. But because of the way semver works on Node, the only version we could release was `0.0.17` (`0.0.13.1` is not a thing). And it should be clear that releasing an higher version number of a package, that contains older code, is a very bad situation.
+
+To address this, in the end with the releases team we had to coordinate a triple version number change for `react-native-codegen` so that it would reshape into:
+
+- `0.68-stable` uses `0.0.17`
+- `0.69-stable` uses `0.69.x`
+- `main` uses `0.70.x`
+
+And all of this had to be done manually, and after each version bump for this package, a new version of the 0.68 and 0.69 had to be released with the updated dependency on the new appropriate version.
+
+This proposal aims to address this problem so that it won't have to be dealt with in the future.
+
+## Detailed design
+
+### The proposal
+
+This proposal will seem deceptively simple: to decouple the root level of the package from being both root and `react-native`, to just be the former, and move `react-native` as a package in the `packages/` folder.
+
+The graph here shows in a bit more detail the end goal, along with a few more details and side-changes that are needed to fully address the current situation:
+
+![proposal for the react-native GitHub codebase](/proposals/assets/0006-proposal.png "Proposal")
+
+Reiterating the changes proposed in the graph:
+
+- step 1 is to move the `react-native` specific code in its own folder within the `packages/` folder. This is the most error prone step to handle, as many paths and variables might have to be modified (for ex. in the CI configurations) to accommodate for this change.
+- step 2 is to **rename** some of other packages as shown above, and in this table:
```

Comment:
Would we want to also standardize naming of the `react-native/packages/*` subfolders at the same time?

Reply:
added a separate table with subfolders name migration, and updated the proposal graph to match 👍